core/provider/provider.go

```diff
@@ -31,12 +31,42 @@ type Task struct {
 	PostStop func(context.Context, *Task) error
 }
 
+type TaskI interface {
+	Initialize(context.Context) error
+	Start(context.Context) error
+	Stop(context.Context) error
+	Destroy(context.Context) error
+
+	GetStatus(context.Context) (TaskStatus, error)
+
+	Modify(context.Context, TaskDefinition) error
+
+	WriteFile(context.Context, string, []byte) error
+	ReadFile(context.Context, string) ([]byte, error)
+	DownloadDir(context.Context, string, string) error
+
+	GetIP(context.Context) (string, error)
+	GetExternalAddress(context.Context, string) (string, error)
+
+	RunCommand(context.Context, []string) (string, string, int, error)
+	RunCommandWhileStopped(context.Context, []string) (string, string, int, error)
+}
+
+type ProviderI interface {
```

Comment:
The existence of `ProviderI` and `Provider` which are both interfaces is pretty confusing here